File: src/lib/shell-read-files.ts
const FILE_READING_COMMANDS = new Set([
	"cat",
	"head",
	"tail",
	"nl",
	"less",
	"more",
	"bat",
	"tac",
	"sed",
]);

const OPTION_VALUE_FLAGS = new Set([
	"--lines",
	"-c",
	"--bytes",
	"-s",
	"--style",
	"--theme",
	"--language",
]);

export function extractShellReadFilePaths(command: string): string[] {
	const inner = unwrapShell(command);
	if (!inner || hasOutputRedirect(inner)) return [];

	const paths: string[] = [];
	for (const segment of splitShellSegments(inner)) {
		const firstPipelineSegment = splitUnquoted(segment, "|")[0]?.trim() ?? "";
		if (!firstPipelineSegment) continue;

		const tokens = tokenizeShell(firstPipelineSegment);
		const commandIndex = tokens.findIndex(
			(token) => token && (!token.includes("=") || token.startsWith("-")),
		);
		if (commandIndex < 0) continue;

		const commandName = basenameToken(tokens[commandIndex]!);
		if (!FILE_READING_COMMANDS.has(commandName)) continue;

		for (const path of extractPathTokens(
			commandName,
			tokens.slice(commandIndex + 1),
		)) {
			if (!paths.includes(path)) {
				paths.push(path);
			}
		}
	}
	return paths;
}

function extractPathTokens(commandName: string, args: string[]): string[] {
	const positional: string[] = [];
	for (let index = 0; index < args.length; index += 1) {
		const token = args[index]!;
		if (!token || token === "--") continue;
		if (
			OPTION_VALUE_FLAGS.has(token) ||
			(commandName !== "sed" && token === "-n")
		) {
			index += 1;
			continue;
		}
		if (token.startsWith("-")) continue;
		positional.push(token);
	}

	if (commandName === "sed") {
		return positional.slice(1).filter(looksLikePath);
	}
	return positional.filter(looksLikePath);
}

function looksLikePath(token: string): boolean {
	if (!token || token === "-") return false;
	if (/^\d+(,\d+)?p$/.test(token)) return false;
	return (
		token.includes("/") ||
		token.includes(".") ||
		token.startsWith("~") ||
		token.startsWith("$")
	);
}

function unwrapShell(command: string): string {
	const trimmed = command.trim();
	const base = trimmed.split(/\s+/, 1)[0]?.split("/").pop() ?? "";
	if (!["sh", "bash", "zsh", "fish", "dash"].includes(base)) {
		return trimmed;
	}

	let rest = trimmed.slice(trimmed.indexOf(base) + base.length).trimStart();
	while (rest.startsWith("-")) {
		const match = rest.match(/^\S+/);
		if (!match) break;
		rest = rest.slice(match[0].length).trimStart();
	}

	if (
		(rest.startsWith('"') && rest.endsWith('"')) ||
		(rest.startsWith("'") && rest.endsWith("'"))
	) {
		return rest.slice(1, -1);
	}
	return rest;
}

function hasOutputRedirect(command: string): boolean {
	let quote: string | null = null;
	for (let index = 0; index < command.length; index += 1) {
		const char = command[index]!;
		if (quote) {
			if (char === "\\" && quote === '"') {
				index += 1;
			} else if (char === quote) {
				quote = null;
			}
			continue;
		}
		if (char === "'" || char === '"') {
			quote = char;
		} else if (char === ">") {
			return true;
		}
	}
	return false;
}

function splitShellSegments(command: string): string[] {
	const segments: string[] = [];
	let start = 0;
	let quote: string | null = null;

	for (let index = 0; index < command.length; index += 1) {
		const char = command[index]!;
		if (quote) {
			if (char === "\\" && quote === '"') {
				index += 1;
			} else if (char === quote) {
				quote = null;
			}
			continue;
		}
		if (char === "'" || char === '"') {
			quote = char;
			continue;
		}
		if (char === ";" || (char === "&" && command[index + 1] === "&")) {
			segments.push(command.slice(start, index));
			if (char === "&") index += 1;
			start = index + 1;
		}
	}
	segments.push(command.slice(start));
	return segments;
}

function splitUnquoted(command: string, separator: string): string[] {
	const parts: string[] = [];
	let start = 0;
	let quote: string | null = null;

	for (let index = 0; index < command.length; index += 1) {
		const char = command[index]!;
		if (quote) {
			if (char === "\\" && quote === '"') {
				index += 1;
			} else if (char === quote) {
				quote = null;
			}
			continue;
		}
		if (char === "'" || char === '"') {
			quote = char;
			continue;
		}
		if (char === separator) {
			parts.push(command.slice(start, index));
			start = index + 1;
		}
	}
	parts.push(command.slice(start));
	return parts;
}

function tokenizeShell(command: string): string[] {
	const tokens: string[] = [];
	let current = "";
	let quote: string | null = null;

	for (let index = 0; index < command.length; index += 1) {
		const char = command[index]!;
		if (quote) {
			if (char === "\\" && quote === '"' && index + 1 < command.length) {
				index += 1;
				current += command[index]!;
			} else if (char === quote) {
				quote = null;
			} else {
				current += char;
			}
			continue;
		}
		if (char === "'" || char === '"') {
			quote = char;
			continue;
		}
		if (/\s/.test(char)) {
			if (current) {
				tokens.push(current);
				current = "";
			}
			continue;
		}
		if (char === "\\" && index + 1 < command.length) {
			index += 1;
			current += command[index]!;
			continue;
		}
		current += char;
	}
	if (current) tokens.push(current);
	return tokens;
}

function basenameToken(token: string): string {
	return token.split("/").pop() ?? token;
}

File: src-tauri/src/pipeline/shell_read_files.rs
pub fn shell_read_file_paths(command: &str) -> Vec<String> {
    let inner = unwrap_shell(command);
    if inner.contains('>') {
        return Vec::new();
    }

    let mut paths = Vec::new();
    for segment in inner.split("&&").flat_map(|part| part.split(';')) {
        let first_pipeline = segment.split('|').next().unwrap_or("").trim();
        if first_pipeline.is_empty() {
            continue;
        }
        let tokens = shell_tokens(first_pipeline);
        let Some(command_index) = tokens
            .iter()
            .position(|token| !token.contains('=') || token.starts_with('-'))
        else {
            continue;
        };
        let command_name = tokens[command_index]
            .rsplit('/')
            .next()
            .unwrap_or(&tokens[command_index]);
        if !matches!(
            command_name,
            "cat" | "head" | "tail" | "nl" | "less" | "more" | "bat" | "tac" | "sed"
        ) {
            continue;
        }

        let mut positional = Vec::new();
        let mut index = command_index + 1;
        while index < tokens.len() {
            let token = &tokens[index];
            if matches!(
                token.as_str(),
                "--lines" | "-c" | "--bytes" | "-s" | "--style" | "--theme" | "--language"
            ) || (command_name != "sed" && token == "-n")
            {
                index += 2;
                continue;
            }
            if token == "--" || token.starts_with('-') {
                index += 1;
                continue;
            }
            positional.push(token.clone());
            index += 1;
        }

        let path_candidates: Vec<String> = if command_name == "sed" {
            positional.into_iter().skip(1).collect()
        } else {
            positional
        };
        for path in path_candidates
            .into_iter()
            .filter(|path| looks_like_path(path))
        {
            if !paths.contains(&path) {
                paths.push(path);
            }
        }
    }
    paths
}

pub fn file_name(path: &str) -> &str {
    path.rsplit(['/', '\\'])
        .find(|segment| !segment.is_empty())
        .unwrap_or(path)
}

fn unwrap_shell(command: &str) -> &str {
    let trimmed = command.trim();
    let first_end = trimmed.find(char::is_whitespace).unwrap_or(trimmed.len());
    let base = trimmed[..first_end]
        .rsplit('/')
        .next()
        .unwrap_or(&trimmed[..first_end]);
    if !matches!(base, "sh" | "bash" | "zsh" | "fish" | "dash") {
        return trimmed;
    }

    let mut rest = trimmed[first_end..].trim_start();
    while rest.starts_with('-') {
        let end = rest.find(char::is_whitespace).unwrap_or(rest.len());
        rest = rest[end..].trim_start();
    }

    let bytes = rest.as_bytes();
    if bytes.len() >= 2 {
        let (first, last) = (bytes[0], bytes[bytes.len() - 1]);
        if (first == b'"' && last == b'"') || (first == b'\'' && last == b'\'') {
            return &rest[1..rest.len() - 1];
        }
    }
    rest
}

fn shell_tokens(command: &str) -> Vec<String> {
    let mut tokens = Vec::new();
    let mut current = String::new();
    let mut quote: Option<char> = None;
    let mut chars = command.chars().peekable();

    while let Some(ch) = chars.next() {
        if let Some(quote_char) = quote {
            if ch == '\\' && quote_char == '"' {
                if let Some(next) = chars.next() {
                    current.push(next);
                }
            } else if ch == quote_char {
                quote = None;
            } else {
                current.push(ch);
            }
            continue;
        }

        if ch == '\'' || ch == '"' {
            quote = Some(ch);
        } else if ch.is_whitespace() {
            if !current.is_empty() {
                tokens.push(std::mem::take(&mut current));
            }
        } else if ch == '\\' {
            if let Some(next) = chars.next() {
                current.push(next);
            }
        } else {
            current.push(ch);
        }
    }
    if !current.is_empty() {
        tokens.push(current);
    }
    tokens
}

fn looks_like_path(token: &str) -> bool {
    if token.is_empty() || token == "-" {
        return false;
    }
    token.contains('/') || token.contains('.') || token.starts_with('~') || token.starts_with('$')
}
